### PR TITLE
[Projects] Adding `allow_cross_project` flag for creating project from template

### PR DESCRIPTION
**Pull request by roei3000b**
**Breaking change in this PR**
We change the way the following functions behave:

1. get_or_create_project
2. load_project

From now,  if the loaded project name(from yaml file) != provided name we will raise Exception and the regular flow of the code will break.

For example, following snippet that worked until this change: 
```
import mlrun
project = mlrun.get_or_create_project(
            "run-with-source-and-auto-build"
        )

# do stuff with project
```

After this change, if the project name from the context != "run-with-source-and-auto-build" it will break this code.
There are 3 ways for fixing it:

1. Set the `allow_cross_project=True` when loading the project.
2. Delete the existing project yaml, or ensure its name is equal to the provided name(run-with-source-and-auto-build)
3. Use different project context dir.

https://iguazio.atlassian.net/browse/ML-6300

### CODE REVIEW

File: docs/projects/project-setup.md [hunk 1]
```diff
@@ -215,4 +215,22 @@ train_function = project.set_function(
 )
 train_function.with_limits(gpus=gpus, cpu=cpu, mem=mem)
 train_function.save()
-```
\ No newline at end of file
+```
+
+### Loading project from template
```
Comment (jillnogold) [inline docs/projects/project-setup.md]: ### Loading a project from a template
You can load a project from a template only if you make one of these changes:
- Set the `allow_cross_project` flag = True and change the name of the project.
- Change the name in the yaml file or delete the file.
- Change the context dir.

File: mlrun/projects/project.py [hunk 2]
```diff
@@ -207,14 +207,14 @@ def setup(project):
                 "Unsupported option, cannot use subpath argument with project templates"
             )
         if from_template.endswith(".yaml"):
-            project = _load_project_file(from_template, name, secrets)
+            project = _load_project_file(from_template, name, secrets, True)
```
Comment (liranbg) [inline mlrun/projects/project.py]:
```suggestion
            project = _load_project_file(from_template, name, secrets, allow_cross_project True)
```
consistency

File: mlrun/projects/project.py [hunk 3]
```diff
@@ -342,6 +343,7 @@ def setup(project):
     :param save:            whether to save the created project and artifact in the DB
     :param sync_functions:  sync the project's functions into the project object (will be saved to the DB if save=True)
     :param parameters:      key/value pairs to add to the project.spec.params
+    :param allow_cross_project: if True, allow loading a project with a different name than the provided name
```
Comment (liranbg) [inline mlrun/projects/project.py]:
```suggestion
    :param allow_cross_project: if True, override the loaded project name. This flag ensures awareness of loading an existing project yaml as a baseline for a new project with a different name
```

File: mlrun/projects/project.py [hunk 4]
```diff
@@ -686,19 +691,28 @@ def _delete_project_from_db(project_name, secrets, deletion_strategy):
...
 
-def _project_instance_from_struct(struct, name):
-    struct.setdefault("metadata", {})["name"] = name or struct.get("metadata", {}).get(
-        "name", ""
-    )
+def _project_instance_from_struct(struct, name, allow_cross_project):
+    name_from_struct = struct.get("metadata", {}).get("name", "")
+    if name and name_from_struct and name_from_struct != name:
+        if allow_cross_project:
+            logger.warn(f"The yaml file was generated for {name_from_struct} project")
```
Comment (liranbg) [inline mlrun/projects/project.py]:
```suggestion
            logger.warn("Project name is different than specified on its project yaml. Overriding.", existing_name=name_from_struct, overriding_name=name)
```

File: mlrun/projects/project.py [hunk 5]
```diff
@@ -686,19 +691,28 @@ def _delete_project_from_db(project_name, secrets, deletion_strategy):
...
+        else:
+            raise ValueError(
+                f"project name mismatch, {name_from_struct} != {name}, please do one of the following:\n"
+                "1. Use the allow_cross_project flag.\n"
```
Comment (liranbg) [inline mlrun/projects/project.py]:
```suggestion
                "1. Set the `allow_cross_project=True` when loading the project.\n"
```

File: mlrun/projects/project.py [hunk 6]
```diff
@@ -686,19 +691,28 @@ def _delete_project_from_db(project_name, secrets, deletion_strategy):
...
+                "2. The yaml file needs to be deleted or edited to change the name in it.\n"
```
Comment (liranbg) [inline mlrun/projects/project.py]:
```suggestion
                f"2. Delete the existing project yaml, or ensure its name is equal to {name}"\n
```

File: mlrun/projects/project.py [hunk 7]
```diff
@@ -686,19 +691,28 @@ def _delete_project_from_db(project_name, secrets, deletion_strategy):
...
+                "3. The context dir needs to be changed."
```
Comment (liranbg) [inline mlrun/projects/project.py]:
```suggestion
                "3. Use different project context dir."
```

File: mlrun/projects/project.py [hunk 8]
```diff
@@ -1814,10 +1828,12 @@ def reload(self, sync=False, context=None) -> "MlrunProject":
         """
         context = context or self.spec.context
         if context:
-            project = _load_project_dir(context, self.metadata.name, self.spec.subpath)
+            project = _load_project_dir(
+                context, self.metadata.name, self.spec.subpath, False
```
Comment (liranbg) [inline mlrun/projects/project.py]:
```suggestion
                context, self.metadata.name, self.spec.subpath, allow_cross_project=False
```

File: mlrun/projects/project.py [hunk 9]
```diff
@@ -1814,10 +1828,12 @@ def reload(self, sync=False, context=None) -> "MlrunProject":
...
+            )
         else:
             project = _load_project_file(
-                self.spec.origin_url, self.metadata.name, self._secrets
+                self.spec.origin_url, self.metadata.name, self._secrets, False
```
Comment (liranbg) [inline mlrun/projects/project.py]:
```suggestion
                self.spec.origin_url, self.metadata.name, self._secrets, allow_cross_project=False
```

**Review by liranbg — CHANGES_REQUESTED** [review]
some wordings and consistency

**Review by liranbg — APPROVED** [review]

**Pull request merged**